runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -2904,281 +2904,303 @@ J9::CodeGenerator::populateOSRBuffer()
 
-   int32_t missedSite = -1;
+         TR_OpaqueMethodBlock *ramMethod = (*info)->_method;
 
-   if (self()->comp()->compileRelocatableCode())
-      {
-      uint32_t inlinedCallSize = self()->comp()->getNumInlinedCallSites();
+         int32_t siteIndex = -1;
 
-      // Create temporary hashtable for ordering AOT guard relocations
-      int32_t counter = inlinedCallSize;
-      TR_InlinedSiteHastTableEntry *orderedInlinedSiteListTable;
-      if (inlinedCallSize > 0)
-         {
-         orderedInlinedSiteListTable= (TR_InlinedSiteHastTableEntry*)self()->comp()->trMemory()->allocateMemory(sizeof(TR_InlinedSiteHastTableEntry) * inlinedCallSize, heapAlloc);
-         memset(orderedInlinedSiteListTable, 0, sizeof(TR_InlinedSiteHastTableEntry)*inlinedCallSize);
+         if (ramMethod != cg->comp()->getCurrentMethod()->getPersistentIdentifier()) // && info->_reloKind != TR_ValidateArbitraryClass)
+            {
+            int32_t i;
+            for (i = 0; i < cg->comp()->getNumInlinedCallSites(); i++)
+               {
+               TR_InlinedCallSite &ics = cg->comp()->getInlinedCallSite(i);
+               TR_OpaqueMethodBlock *inlinedMethod = fej9->getInlinedCallSiteMethod(&ics);
+
+               traceMsg(cg->comp(), "\tinline site %d inlined method %p\n", i, inlinedMethod);
+               if (ramMethod == inlinedMethod)
+                  {
+                  traceMsg(cg->comp(), "\t\tmatch!\n");
+                  siteIndex = i;
+                  break;
+                  }
+               }
+
+            if (i >= (int32_t) cg->comp()->getNumInlinedCallSites())
+               {
+               // this assumption isn't associated with a method directly in the compilation
+               // so we can't use a constant pool approach to validate: transform into TR_ValidateArbitraryClass
+               // kind of overkill for TR_ValidateStaticField, but still correct
+               (*info)->_reloKind = TR_ValidateArbitraryClass;
+               siteIndex = -1;   // invalidate main compiled method
+               traceMsg(cg->comp(), "\ttransformed into TR_ValidateArbitraryClass\n");
+               }
+            }
+
+         traceMsg(cg->comp(), "Found inlined site %d\n", siteIndex);
+
+         TR_ASSERT(siteIndex < (int32_t) cg->comp()->getNumInlinedCallSites(), "did not find AOTClassInfo %p method in inlined site table", *info);
+
+         cg->addExternalRelocation(new (cg->trHeapMemory()) TR::ExternalRelocation(NULL,
+                                                                          (uint8_t *)(intptr_t)siteIndex,
+                                                                          (uint8_t *)(*info),
+                                                                          (*info)->_reloKind, cg),
+                                                                          __FILE__, __LINE__, NULL);
          }
-      else
-         orderedInlinedSiteListTable = NULL;
+      }
```

Comment:
I haven't changed any of the logic that happens in here; I just moved it to its own function so that it's just easier to read the code. However, I believe something like that happens when writing the header in J9AheadOfTimeCompile.cpp; I think it's called `findCorrectInlinedSiteIndex`.